### PR TITLE
coverage: use LLVM and bazel native coverage

### PR DESCRIPTION
**Pull request by lizan**
Signed-off-by: Lizan Zhou <lizan@tetrate.io>

Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:

### CODE REVIEW

**Comment by lizan** [comment]
This is ready, still using single binary though LLVM and bazel native coverage. CI fails because coverage is below threshold (96.94%). The report is here and should be compared to coverage report from #7566: https://248596-65214191-gh.circle-artifacts.com/0/coverage/index.html

**Comment by mattklein123** [comment]
@lizan agreed let's drop the % required for now and ship it. Looking through there is legit coverage missing so maybe we can raise it again one day.

Q: Which project(s) would we file issues against to get the ability to exclude lines back? It seems like something that shouldn't be that hard to do and maybe we could list as a starter project somewhere. (Either way let's open an issue to track further coverage improvements.)

/wait

**Comment by lizan** [comment]
@mattklein123 With #7568 I think 97.0% is the threshold we can use for now. It is 0.5% down from currently, which is the gap of excluded lines.

> Q: Which project(s) would we file issues against to get the ability to exclude lines back? It seems like something that shouldn't be that hard to do and maybe we could list as a starter project somewhere. (Either way let's open an issue to track further coverage improvements.)

It is `llvm-cov`. Yeah from a quick glance it shouldn't be too hard to do.

File: ci/do_ci.sh [hunk 1]
```diff
@@ -241,22 +241,11 @@ elif [[ "$CI_TARGET" == "bazel.api" ]]; then
     @envoy_api//tools:tap2pcap_test
   exit 0
 elif [[ "$CI_TARGET" == "bazel.coverage" ]]; then
-  setup_gcc_toolchain
+  setup_clang_toolchain
   echo "bazel coverage build with tests ${TEST_TARGETS}"
 
-  # gcovr is a pain to run with `bazel run`, so package it up into a
-  # relocatable and hermetic-ish .par file.
-  bazel build --python_version=PY2 @com_github_gcovr_gcovr//:gcovr.par
```
Comment (mattklein123) [inline ci/do_ci.sh]: Can we kill gcov as a dep?

Comment (lizan) [reply ci/do_ci.sh]: Ah yes, nice catch.

**Comment by lizan** [comment]
There is already one in LLVM actually: https://bugs.llvm.org/show_bug.cgi?id=34277

File: test/run_envoy_bazel_coverage.sh [hunk 2]
```diff
@@ -31,97 +21,43 @@ else
...
-
-echo "Cleaning .gcda/.gcov from previous coverage runs..."
-NUM_PREVIOUS_GCOV_FILES=0
-for f in $(find -L "${GCOVR_DIR}" -name "*.gcda" -o -name "*.gcov")
-do
-  rm -f "${f}"
-  let NUM_PREVIOUS_GCOV_FILES=NUM_PREVIOUS_GCOV_FILES+1
-done
-echo "Cleanup completed. ${NUM_PREVIOUS_GCOV_FILES} files deleted."
+NO_GCOV=1 "${SCRIPT_DIR}"/coverage/gen_build.sh ${COVERAGE_TARGETS}
```
Comment (htuch) [inline test/run_envoy_bazel_coverage.sh]: Maybe just remove this conditioning from `gen_build.sh`

File: test/run_envoy_bazel_coverage.sh [hunk 3]
```diff
@@ -31,97 +21,43 @@ else
...
-# stats. The #foo# pattern is because gcov produces files such as
-# bazel-out#local-fastbuild#bin#external#spdlog_git#_virtual_includes#spdlog#spdlog#details#pattern_formatter_impl.h.gcov.
-# To find these while modifying this regex, perform a gcov run with -k set.
-[[ -z "${GCOVR_EXCLUDE_REGEX}" ]] && GCOVR_EXCLUDE_REGEX=".*pb\\..*|test#.*|.*#test#.*|external#.*|.*#external#.*|.*#prebuilt#.*|.*#config_validation#.*|.*#chromium_url#.*"
-[[ -z "${GCOVR_EXCLUDE_DIR}" ]] && GCOVR_EXCLUDE_DIR=".*/external/.*"
+BAZEL_USE_LLVM_NATIVE_COVERAGE=1 GCOV=llvm-profdata bazel coverage ${BAZEL_BUILD_OPTIONS} \
+    -c fastbuild --copt=-DNDEBUG --instrumentation_filter=//source/...,//include/... \
+    --test_timeout=2000 --cxxopt="-DENVOY_CONFIG_COVERAGE=1" --test_output=streamed \
+    --test_arg="--log-path /dev/null" --test_arg="-l trace" --test_env=HEAPCHECK= \
+    --test_filter='-QuicPlatformTest.QuicStackTraceTest' //test/coverage:coverage_tests
```
Comment (htuch) [inline test/run_envoy_bazel_coverage.sh]: Why is this test filtered? Can you add a comment or TODO?

**Comment by htuch** [comment]
@asraa FYI, this should change how coverage is done in OSS Envoy and it should make it faster to iterate around the nghttp2 coverage effort.

File: test/run_envoy_bazel_coverage.sh (hunk 2 above)
Comment (lizan) [reply test/run_envoy_bazel_coverage.sh]: sure, removed as well as gcc_only_test.

File: test/run_envoy_bazel_coverage.sh (hunk 3 above)
Comment (lizan) [reply test/run_envoy_bazel_coverage.sh]: actually this worked in single binary so just removed filter, let's see if CI is happy with this.

**Pull request merged**